File: m02_ros2_communication/m02_ros2_with_cpp/src/saying_hi.cpp
/**
 * File: saying_hi.pp
 * 
 * Description: A simple node that prints the node saying hi with a name, that
 * can be modified by changing the value of the str param called 'your name'.
*/

// ------------------------------ CPP Standard Library -----------------------

#include <chrono>      // For flexible collection of types that tracks time with
                       // varying degrees of precision.

#include <functional>  // Function objects library, related with hash functions,
                       // those function with outputs that depends only on the
                       // input.

#include <string>      // Management and usage of strings.

// ------------------------------ rclcpp headers ------------------------------

#include <rclcpp/rclcpp.hpp> // Proper include for using rclcpp

// -------------------------- Implementation of namespaces --------------------
using namespace std::chrono_literals;

// ------------------------- Simple Displayer of Hi Class ---------------------

/**
 * Class that display a node saying hi and appending a name linked to a param,
 * that can be modified in the terminal or by other node.
*/
class SayingHi : public rclcpp::Node
{
  // Public interface
  public:
    /**
     * Constructo call that initialize a node with a given name (using parent 
     * interface), instance a parameter (with its own descriptor), and instnace
     * a timer with a fixed rate linked to a callback for pinting the info.
    */
    SayingHi() : Node("saying_your_name")
    {
      auto param_desc = rcl_interfaces::msg::ParameterDescriptor{};
      param_desc.description = 
        "Put your name here, so you will be called by it!";
      this->declare_parameter("your_name", "ROS User");

      timer_ = this->create_wall_timer(
        5000ms, std::bind(&SayingHi::say_callback, this));
    }

    /**
     * Callback that will be used when the timer indicate to print a new message
     * on the terminal.
    */
    void say_callback()
    {
      // Get param value
      std::string change_param = this->get_parameter("your_name").as_string();

      // Display info
      RCLCPP_INFO(this->get_logger(), "Hi! I am a node, and you are %s.",
        change_param.c_str());

      // Update the param value with the original one
      std::vector<rclcpp::Parameter> new_params{
        rclcpp::Parameter("your_name", "ROS User")};
      this->set_parameters(new_params);
    }

  private:
    // Instance a timer object using a shared pointer
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv)
{
  // Initialize ROS Client Library for CPP
  rclcpp::init(argc, argv);

  // Spin (loop) for multiple prints
  rclcpp::spin(std::make_shared<SayingHi>());

  // CLose and shutdown before final return
  rclcpp::shutdown();
  return 0;
}

File: m02_ros2_communication/m02_ros2_with_py/m02_ros2_with_py/saying_hi.py
# !/usr/bin/env python3

# ------------------------------- ROS2 Dependencies ---------------------------
import rclpy
import rclpy.node
from rcl_interfaces.msg import ParameterDescriptor

#-------------------------- Printer Class related with Params -----------------
class SayingHi(rclpy.node.Node):
    """
    Class oriented to print message of hi, as a presentation of itself and by
    adding a name.
    
    Attributes
    ---
    timer - Timer object
        Time counter that links a callback when the period is completed.
        
    Methods
    ---
        saying_callback()
            Print a little presentation with a hi, and the name indicated in
            the params associated.
    """
    
    def __init__(self):
        """
        Constructor that initialize the node (by passing a name), instance a
        param with its description and then declares it to make it accesible
        for other nodes.
        """
        super().__init__("saying_your_name")
        param_desc = ParameterDescriptor(description = "Put here your name")
        self.declare_parameter('your_name', 'ROS User', param_desc)
        self.timer = self.create_timer(1, self.saying_callback)
        
    def saying_callback(self):
        """
        Callback related with the printing of the message, saying hi to the
        user, and updating its value to the original one if a change is 
        received.
        """
        change_param = self.get_parameter('your_name').get_parameter_value().string_value
        self.get_logger().info('Hi! I am node and you are %s' % str(change_param))
        new_param = rclpy.parameter.Parameter('your_name', rclpy.Parameter.Type.STRING, 'ROS User')
        self.set_parameters([new_param])
        
# --------------------------- Implementation ----------------------------------
def main():
    """
    Simple program that iterates a saying hi object actions to wave to the 
    user.
    """
    
    # Initialize ROS LCient Library for Python
    rclpy.init()
    
    # Instance object to print message to the user
    node = SayingHi()
    
    # Spin node
    rclpy.spin(node)
    
    # It is a good practice to destroy the node and shutdown rclpy, but if it
    # forgotten, the garbage collector will take care of it (in the future)
    node.destroy_node()
    rclpy.shutdown()



if __name__ == '__main__':
    main()